src/dhtproto/client/legacy/internal/helper/Handshake.d

```diff
@@ -0,0 +1,211 @@
+
+        Task that suspends until at least one node is connected
+
+    ***************************************************************************/
+
+    private class OneDhtNodeConnected : Task
+    {
+        import swarm.Const : NodeItem;
+
+        /***********************************************************************
+
+            Boolean flag that will be set to `true` once at least one DHT node
+            has been reached by the handshake
+
+        ***********************************************************************/
+
+        private bool at_least_one_node_connected = false;
+
+        /***********************************************************************
+
+            Delegate to be called by the handshake each time an individual
+            node connects
+
+        ***********************************************************************/
+
+        private void dg (NodeItem item)
+        {
+            if (!this.at_least_one_node_connected)
+            {
+                this.at_least_one_node_connected = true;
+                if (this.suspended()) this.resume();
+            }
+
+            this.outer.log.trace("Connected to DHT node {}:{}",
+                                 item.Address, item.Port);
+        }
+
+        /***********************************************************************
+
+            Runs the task
+
+        ***********************************************************************/
+
+        protected override void run ()
+        {
+            if (!this.at_least_one_node_connected) this.suspend();
+        }
+    }
+
+
+    /***************************************************************************
+
+        Logger for the handshake
+
+    ***************************************************************************/
+
+    private Logger log;
+
+
+    /***************************************************************************
+
+        Instance of task that suspends until all nodes are connected
+
+    ***************************************************************************/
+
+    private AllDhtNodesConnected all_nodes_connected;
+
+
+    /***************************************************************************
+
+        Instance of task that suspends until at least one node is connected
+
+    ***************************************************************************/
+
+    private OneDhtNodeConnected one_node_connected;
+
+
+    /***************************************************************************
+
+        Instance of class that launches the actual handshake, retrying when
+        it fails until all nodes are connected
+
+    ***************************************************************************/
+
+    private RetryHandshake retry_handshake;
+
+
+    /***************************************************************************
+
+        Constructor
+
+        Params:
+            epoll = epoll select dispatcher via which to schedule handshake
```

Comment:
If this whole thing is based on tasks, is there any reason to accept an epoll selector here? You could just use `theScheduler.epoll`, internally.

Reply:
I didn't want to hardcode an unnecessary dependency on `theScheduler`, on the grounds that the user might have a different setup (e.g. custom `Scheduler` instance instead of `theScheduler`).  But that might be over-engineering.

Reply:
> e.g. custom Scheduler instance instead of theScheduler

This is not supported by ocean